### PR TITLE
feat: add uint2str builtin

### PR DESCRIPTION
**Pull request by charles-cooper**
### What I did
implement https://github.com/vyperlang/vyper/issues/2703

### Cute Animal Picture

image

### CODE REVIEW

**Comment by fubuloubu** [comment]
Can also add implementation to here: https://github.com/vyperlang/vyper/blob/master/examples/tokens/ERC721.vy

```vyper
baseURL: String[53]  # ipfs://Qm....

@view
@external
def tokenURL(tokenId: uint256) -> String[132]:
    return concat(self.baseURL, uint2str(tokenId))
```

**Comment by skellet0r** [comment]
I can already sense the usage of this with a constant ...
`uint2str(3000)` 🤦‍♀️

Would adding an `evaluate` method for compile time check be good here.

**Review by skellet0r — APPROVED** [review]

File: examples/tokens/ERC721.vy [hunk 1]
```diff
@@ -82,6 +84,7 @@ def __init__():
     @dev Contract constructor.
     """
     self.minter = msg.sender
+    self.baseURL = "some url"
```
Comment (fubuloubu) [inline examples/tokens/ERC721.vy:87]:
```suggestion
    self.baseURL = "https://api.babby.xyz/metadata/"
```

**Comment by charles-cooper** [comment]
> I can already sense the usage of this with a constant ... `uint2str(3000)` woman_facepalming
> 
> Would adding an `evaluate` method for compile time check be good here.

:facepalm: `"3000"` is even fewer characters than `uint2str(3000)`. i mean maybe we should even block `uint2str` on literals

**Comment by charles-cooper** [comment]
> facepalm `"3000"` is even fewer characters than `uint2str(3000)`. i mean maybe we should even block `uint2str` on literals

never mind, it makes sense -- `3000` could be a result of constant propagation

File: vyper/builtin_functions/functions.py [hunk 2]
```diff
@@ -1807,6 +1807,64 @@ class Max(_MinMax):
     _opcode = "gt"
 
 
+class UintToStr(_SimpleBuiltinFunction):
+    _id = "uint2str"
```
[the remark this answers is not in the document]
Comment (charles-cooper) [reply vyper/builtin_functions/functions.py:1811]: how about `str()` after python?

File: vyper/builtin_functions/functions.py [hunk 3]
```diff
@@ -1807,6 +1807,64 @@ class Max(_MinMax):
...
+    _inputs = [("x", Uint256Definition())]  # should allow any uint?
```
[the remark this answers is not in the document]
Comment (charles-cooper) [reply vyper/builtin_functions/functions.py:1888]: i don't know. in the interest of simplicity, i think we should leave it as is for now, and then maybe add more types if people complain.

[the remark this answers is not in the document]
Comment (skellet0r) [reply vyper/builtin_functions/functions.py:1888]: Yeah, plus can always do some combination of converting and using abs built in.

For signed:

`str(abs(singed_int))`

**Review by fubuloubu — COMMENTED** [review]
My one misgivings with this PR is that I don't think it would be possible to signal when a type doesn't need the full string size for itself, like when converting `tokenId` for a 10k token mint set

**Comment by charles-cooper** [comment]
i changed the builtin to allow a uint of any width, this also gives fine-grained control over the output string length.

**Comment by charles-cooper** [comment]
switching back to `uint2str`, since we don't want to suggest that we have general `str` implementations for everything. if we start to add more builtins, we can try to settle on a more "generic" name like `str` or `format`.

**Review by fubuloubu — APPROVED** [review]

**Pull request merged**